storage/range.go

```diff
@@ -1543,38 +1570,61 @@ func (r *Range) InternalTruncateLog(batch engine.Engine, ms *proto.MVCCStats, ar
-// expiration will be extended as indicated. Otherwise, all duties
-// required of the range leader are commenced, including gossiping
-// cluster and config info where appropriate and clearing the command
-// queue and timestamp cache.
+// InternalLeaderLease sets the leader lease for this range. The command fails
+// only if the desired start timestamp collides with a previous lease.
+// Otherwise, the start timestamp is wound back to right after the expiration
+// of the previous lease (or zero). After a lease has been set, calls to
+// HasLeaderLease() will return true if this replica is the lease holder and
+// the lease has not yet expired. If this range replica is already the lease
+// holder, the expiration will be extended or shortened as indicated. For a new
+// lease, all duties required of the range leader are commenced, including
+// clearing the command queue and timestamp cache.
 func (r *Range) InternalLeaderLease(batch engine.Engine, ms *proto.MVCCStats, args *proto.InternalLeaderLeaseRequest, reply *proto.InternalLeaderLeaseResponse) {
 	r.Lock()
 	defer r.Unlock()
 
-	// Verify details of new lease request. The start of this lease must
-	// be afer the prior lease's expiration.
 	oldLease := r.getLease()
+	// Verify details of new lease request. The start of this lease must
+	// obviously precede its expiration.
 	if !args.Lease.Start.Less(args.Lease.Expiration) {
-		reply.SetGoError(util.Errorf("lease %d - %d invalid", args.Lease.Start, args.Lease.Expiration))
+		reply.SetGoError(util.Errorf("lease %s - %s invalid",
+			args.Lease.Start, args.Lease.Expiration))
 		return
-	} else if oldLease.RaftNodeID == args.Lease.RaftNodeID {
-		if !oldLease.Start.Less(args.Lease.Start) {
-			reply.SetGoError(util.Errorf("lease renewal %d - %d invalid; prior lease %d - %d",
-				args.Lease.Start, args.Lease.Expiration, oldLease.Start, oldLease.Expiration))
+	}
+
+	// Wind the start timestamp back as far to the previous lease's expiration
+	// as we can. That'll make sure that when multiple leases are requested out
+	// of order at the same replica (after all, they use the request timestamp,
+	// which isn't straight out of our local clock), they all succeed
+	// nevertheless, unless they have a "real" issue with a previous lease.
+	// Example: Assuming no previous lease, one request for [5, 15) followed by
+	// one for [0, 15) would fail without this optimization. With it, the first
+	// request effectively gets the lease for [0, 15), which the second one can
+	// commit again (even extending your own lease is possible; see below).
+	effectiveStart := oldLease.Expiration
+	// If no old lease exists or this is our lease, we don't need to add this
+	// extra tick. Unfortunately relevant in practice, especially in tests which
+	// tend to have manual clocks and run operations at the zero timestamp.
+	if oldLease.RaftNodeID != 0 && oldLease.RaftNodeID != args.Lease.RaftNodeID {
+		effectiveStart = effectiveStart.Add(0, 1)
```

Comment:
Ack, another test-driven if statement. Did you try just setting the manual clock to ts(0, 1) in range_test's testContext?